File: thea-tizen/src/services/ai/ProviderRegistry.ts
/**
 * Provider Registry - Central management of AI providers
 * Mirrors: Shared/Providers/Registry/ProviderRegistry.swift
 */

import type { AIProvider, ProviderCapability } from './AIProvider';
import type { AIModel, ProviderHealth } from '../../types/chat';
import { AnthropicProvider } from './AnthropicProvider';
import { OpenRouterProvider } from './OpenRouterProvider';

/**
 * Singleton registry for managing AI providers
 */
class ProviderRegistryClass {
  private providers: Map<string, AIProvider> = new Map();
  private healthCache: Map<string, ProviderHealth> = new Map();
  private defaultProviderId: string = 'openrouter';

  constructor() {
    // Register default providers
    this.register(new AnthropicProvider());
    this.register(new OpenRouterProvider());
  }

  /**
   * Register a provider
   */
  register(provider: AIProvider): void {
    this.providers.set(provider.id, provider);
  }

  /**
   * Unregister a provider
   */
  unregister(providerId: string): void {
    this.providers.delete(providerId);
    this.healthCache.delete(providerId);
  }

  /**
   * Get a provider by ID
   */
  getProvider(id: string): AIProvider | undefined {
    return this.providers.get(id);
  }

  /**
   * Get the default provider
   */
  get defaultProvider(): AIProvider | undefined {
    return this.providers.get(this.defaultProviderId);
  }

  /**
   * Set the default provider
   */
  setDefaultProvider(providerId: string): void {
    if (this.providers.has(providerId)) {
      this.defaultProviderId = providerId;
    }
  }

  /**
   * Get all registered providers
   */
  get allProviders(): AIProvider[] {
    return Array.from(this.providers.values());
  }

  /**
   * Get only configured providers
   */
  get configuredProviders(): AIProvider[] {
    return this.allProviders.filter(p => p.isConfigured);
  }

  /**
   * Get providers with specific capability
   */
  getProvidersWithCapability(capability: ProviderCapability): AIProvider[] {
    return this.allProviders.filter(p => p.capabilities.has(capability));
  }

  /**
   * Get all models from all providers
   */
  get allModels(): AIModel[] {
    return this.allProviders.flatMap(p => p.supportedModels);
  }

  /**
   * Get models from configured providers only
   */
  get availableModels(): AIModel[] {
    return this.configuredProviders.flatMap(p => p.supportedModels);
  }

  /**
   * Find a model by ID
   */
  findModel(modelId: string): { provider: AIProvider; model: AIModel } | undefined {
    for (const provider of this.allProviders) {
      const model = provider.supportedModels.find(m => m.id === modelId);
      if (model) {
        return { provider, model };
      }
    }
    return undefined;
  }

  /**
   * Get the best available provider (first configured one)
   */
  get bestAvailableProvider(): AIProvider | undefined {
    // Prefer default if configured
    const defaultProv = this.defaultProvider;
    if (defaultProv?.isConfigured) {
      return defaultProv;
    }

    // Return first configured provider
    return this.configuredProviders[0];
  }

  /**
   * Check health of all providers
   */
  async checkAllHealth(): Promise<Map<string, ProviderHealth>> {
    const checks = this.configuredProviders.map(async provider => {
      const health = await provider.checkHealth();
      this.healthCache.set(provider.id, health);
      return [provider.id, health] as const;
    });

    const results = await Promise.all(checks);
    return new Map(results);
  }

  /**
   * Get cached health for a provider
   */
  getHealth(providerId: string): ProviderHealth | undefined {
    return this.healthCache.get(providerId);
  }

  /**
   * Get registry statistics
   */
  getStatistics(): {
    totalProviders: number;
    configuredProviders: number;
    healthyProviders: number;
    totalModels: number;
    availableModels: number;
  } {
    const healthyCount = Array.from(this.healthCache.values()).filter(
      h => h.isHealthy
    ).length;

    return {
      totalProviders: this.providers.size,
      configuredProviders: this.configuredProviders.length,
      healthyProviders: healthyCount,
      totalModels: this.allModels.length,
      availableModels: this.availableModels.length,
    };
  }

  /**
   * Configure a provider with an API key
   */
  configureProvider(providerId: string, apiKey: string): boolean {
    const provider = this.providers.get(providerId);
    if (provider && 'setApiKey' in provider) {
      (provider as { setApiKey: (key: string) => void }).setApiKey(apiKey);
      return true;
    }
    return false;
  }

  /**
   * Load API keys from storage
   */
  loadFromStorage(storage: {
    getItem: (key: string) => string | null;
  }): void {
    const keyMap: Record<string, string> = {
      anthropic: 'thea_api_anthropic',
      openrouter: 'thea_api_openrouter',
      openai: 'thea_api_openai',
      google: 'thea_api_google',
    };

    for (const [providerId, storageKey] of Object.entries(keyMap)) {
      const apiKey = storage.getItem(storageKey);
      if (apiKey) {
        this.configureProvider(providerId, apiKey);
      }
    }
  }
}

// Create singleton instance
const registryInstance = new ProviderRegistryClass();

// Export with static helper methods
export const ProviderRegistry = Object.assign(registryInstance, {
  /** Get singleton instance */
  getInstance: () => registryInstance,
  /** Get the best available configured provider */
  getDefaultProvider: () => registryInstance.bestAvailableProvider,
});

File: Shared/Providers/Registry/ProviderRegistry.swift
// ProviderRegistry.swift
// Thea V2
//
// Central registry for all AI providers

import Foundation
import OSLog

// MARK: - Provider Registry

/// Central registry managing all AI providers
@MainActor
public final class ProviderRegistry: ObservableObject {
    public static let shared = ProviderRegistry()

    private let logger = Logger(subsystem: "com.thea.v2", category: "ProviderRegistry")

    // MARK: - Published State

    @Published public private(set) var providers: [String: any AIProvider] = [:]
    @Published public private(set) var defaultProviderId: String = "openrouter"
    @Published public private(set) var providerHealth: [String: ProviderHealth] = [:]

    // MARK: - Initialization

    private init() {
        // Providers will be registered during app startup
    }

    // MARK: - Registration

    /// Register a provider
    public func register(_ provider: any AIProvider) {
        providers[provider.id] = provider
        logger.info("Registered provider: \(provider.name)")

        // Publish event
        EventBus.shared.publish(StateEvent(
            source: .system,
            component: "ProviderRegistry",
            newState: "registered:\(provider.id)"
        ))
    }

    /// Unregister a provider
    public func unregister(_ providerId: String) {
        providers.removeValue(forKey: providerId)
        providerHealth.removeValue(forKey: providerId)
        logger.info("Unregistered provider: \(providerId)")
    }

    // MARK: - Access

    /// Get a provider by ID
    public func provider(id: String) -> (any AIProvider)? {
        providers[id]
    }

    /// Get the default provider
    public var defaultProvider: (any AIProvider)? {
        providers[defaultProviderId]
    }

    /// Set the default provider
    public func setDefault(_ providerId: String) {
        guard providers[providerId] != nil else {
            logger.warning("Cannot set default to unknown provider: \(providerId)")
            return
        }
        defaultProviderId = providerId

        // Update config
        TheaConfig.shared.ai.defaultProvider = providerId
        TheaConfig.shared.save()

        // Publish event
        EventBus.shared.publish(StateEvent(
            source: .user,
            component: "ProviderRegistry",
            previousState: defaultProviderId,
            newState: providerId,
            reason: "User changed default provider"
        ))
    }

    /// Get all configured providers
    public var configuredProviders: [any AIProvider] {
        providers.values.filter(\.isConfigured)
    }

    /// Get providers with specific capability
    public func providers(with capability: ProviderCapability) -> [any AIProvider] {
        providers.values.filter { $0.capabilities.contains(capability) }
    }

    // MARK: - Models

    /// Get all available models across all configured providers
    public var allModels: [AIModel] {
        configuredProviders.flatMap(\.supportedModels)
    }

    /// Get models for a specific provider
    public func models(for providerId: String) -> [AIModel] {
        providers[providerId]?.supportedModels ?? []
    }

    /// Find a model by ID
    public func findModel(_ modelId: String) -> (provider: any AIProvider, model: AIModel)? {
        for provider in providers.values {
            if let model = provider.supportedModels.first(where: { $0.id == modelId }) {
                return (provider, model)
            }
        }
        return nil
    }

    // MARK: - Health Checks

    /// Check health of all providers
    public func checkAllHealth() async {
        for (id, provider) in providers {
            let health = await provider.checkHealth()
            providerHealth[id] = health

            if !health.isHealthy {
                logger.warning("Provider \(id) unhealthy: \(health.errorMessage ?? "Unknown")")
            }
        }
    }

    /// Check health of a specific provider
    public func checkHealth(for providerId: String) async -> ProviderHealth? {
        guard let provider = providers[providerId] else { return nil }

        let health = await provider.checkHealth()
        providerHealth[providerId] = health

        return health
    }

    /// Get provider health status
    public func health(for providerId: String) -> ProviderHealth? {
        providerHealth[providerId]
    }

    // MARK: - Chat Helpers

    /// Send a chat request using the best available provider
    public func chat(
        messages: [ChatMessage],
        model: String? = nil,
        options: ChatOptions = .default
    ) async throws -> AsyncThrowingStream<StreamChunk, Error> {
        // Find provider and model
        let (provider, _) = try resolveProviderAndModel(modelId: model)

        let modelId = model ?? TheaConfig.shared.ai.defaultModel

        // Log action
        EventBus.shared.logAction(
            .modelQuery,
            target: modelId,
            parameters: ["provider": provider.id],
            success: true
        )

        return try await provider.chat(
            messages: messages,
            model: modelId,
            options: options
        )
    }

    /// Send a sync chat request
    public func chatSync(
        messages: [ChatMessage],
        model: String? = nil,
        options: ChatOptions = .default
    ) async throws -> ChatResponse {
        let (provider, _) = try resolveProviderAndModel(modelId: model)
        let modelId = model ?? TheaConfig.shared.ai.defaultModel

        return try await provider.chatSync(
            messages: messages,
            model: modelId,
            options: options
        )
    }

    // MARK: - Resolution

    private func resolveProviderAndModel(modelId: String?) throws -> (any AIProvider, AIModel) {
        let targetModelId = modelId ?? TheaConfig.shared.ai.defaultModel

        // First try to find model across providers
        if let found = findModel(targetModelId) {
            guard found.provider.isConfigured else {
                throw ProviderError.notConfigured(provider: found.provider.name)
            }
            return found
        }

        // Try default provider
        if let provider = defaultProvider, provider.isConfigured {
            if let model = provider.supportedModels.first {
                return (provider, model)
            }
        }

        // Try any configured provider
        if let provider = configuredProviders.first,
           let model = provider.supportedModels.first {
            return (provider, model)
        }

        throw ProviderError.notConfigured(provider: "No providers configured")
    }

    // MARK: - Statistics

    public struct RegistryStatistics: Sendable {
        public let totalProviders: Int
        public let configuredProviders: Int
        public let healthyProviders: Int
        public let totalModels: Int
    }

    public func getStatistics() -> RegistryStatistics {
        let healthy = providerHealth.values.filter(\.isHealthy).count

        return RegistryStatistics(
            totalProviders: providers.count,
            configuredProviders: configuredProviders.count,
            healthyProviders: healthy,
            totalModels: allModels.count
        )
    }
}

// MARK: - Provider Registration Extension

public extension ProviderRegistry {
    /// Register all built-in providers
    func registerBuiltInProviders(keychain: KeychainAccess) {
        // This will be called during app initialization
        // Each provider checks keychain for its API key
        logger.info("Provider registration will occur during app startup")
    }
}

// MARK: - V1 Compatibility

public extension ProviderRegistry {
    /// V1 compatibility method
    func getProvider(id: String) -> (any AIProvider)? {
        provider(id: id)
    }

    /// V1 compatibility - all registered providers
    var availableProviders: [any AIProvider] {
        Array(providers.values)
    }
}

// MARK: - Keychain Access Protocol

/// Protocol for accessing stored API keys
public protocol KeychainAccess: Sendable {
    func get(_ key: String) -> String?
    func set(_ value: String, for key: String) throws
    func delete(_ key: String) throws
}
